crates/wasmparser/src/validator/component_types.rs

```diff
@@ -928,20 +1064,106 @@ impl ComponentFuncType {
+
+        let func_ty = sig.to_func_type();
+
+        // Check expected arity: shouldn't have any left-over expected types
+        // that we haven't matched against yet because we stopped pushing types.
+        //
+        // Note that we also delay reporting `len(actual) > len(expected)`
+        // arity-mismatch errors until this point -- even though we can
+        // technically detect them earlier -- because we can craft better error
+        // messages with the context available here.
+        if let Some(e) = sig.params.expected.as_ref() {
+            let actual_len = func_ty.params().len();
+            let expected_len = e.expected.len();
+            if actual_len != expected_len {
+                bail!(
+                    offset,
+                    "type mismatch when checking `core-type` canonical option\n\
+                     arity mismatch: expected {expected_len} parameters, found {actual_len} parameters\n\
+                     expected = {}\n\
+                     actual   = {func_ty}",
+                    types[options.core_type.unwrap()].unwrap_func(),
+                );
+            }
+        }
+        if let Some(e) = sig.results.expected.as_ref() {
+            let actual_len = func_ty.results().len();
+            let expected_len = e.expected.len();
+            if actual_len != expected_len {
+                bail!(
+                    offset,
+                    "type mismatch when checking `core-type` canonical option\n\
+                     arity mismatch: expected {expected_len} results, found {actual_len} results\n\
+                     expected = {}\n\
+                     actual   = {func_ty}",
+                    types[options.core_type.unwrap()].unwrap_func(),
+                );
+            }
+        }
+
+        #[cfg(debug_assertions)]
+        if let Some(expected) = options.core_type {
+            let expected = types[expected].unwrap_func();
+            debug_assert!(
+                Matches::matches(types, &func_ty, expected),
+                "should have validated that the actual lowered type matches the \
+                 expected type in the process of lowering!\n\
+                 \n\
+                 expected = {expected}\n\
+                 actual   = {func_ty}"
+            );
         }
 
+        Ok(func_ty)
+    }
+
+    fn lower_params(
+        &self,
+        types: &TypeList,
+        options: &CanonicalOptions,
+        abi: Abi,
+        offset: usize,
+        sig: &mut LoweredSignature<'_>,
+    ) -> Result<()> {
+        match self.lower_params_flat(types, options, abi, offset, sig) {
+            Ok(()) => Ok(()),
+            Err(LowerError::OutOfCapacity) => self.lower_params_indirect(options, abi, offset, sig),
+            Err(LowerError::TypeMismatch(e)) => self
+                .lower_params_indirect(options, abi, offset, sig)
+                .with_context(|| e.to_string()),
+            Err(LowerError::Other(e)) => Err(e),
+        }
```

Comment:
Could you add some comments to this `match`? I think it'd be good to explicitly comment why the `OutOfCapacity` error is caught here for example.

Additionally I don't understand how `TypeMismatch` is handled here. That's taking what I thought would be a fatal error, trying to catch it, and then adding context if the indirect option fails.

Reply:
We initially try to lower flat, but that can fail if we have too many args to pass flatly, in which case we then try to pass indirect. The core types of the flat vs indirect versions are going to be different, so you can fail to flatten due to a type mismatch (which would otherwise become a out-of-capacity error) because the declared type is expecting to be indirect.